File: iOS/Sources/Data/services/TimelineService.swift
import Foundation

struct TimelineFilterPeriod { let start: Date; let end: Date }

struct TimelineFilterParams {
    var period: TimelineFilterPeriod? = nil
    var categoryIds: [String] = []
    var showArchived: Bool = false
    var accountIds: [String] = []
    var paid: Bool? = nil
    var light: Bool = false
    var search: String? = nil
}

struct TimelineDetailsItem {
    let sourceName: String
    let registry: AccountsRegistry
    let category: Category?
}

class TimelineService {
    private let accounts = AccountsRepository()
    private let categories = CategoriesRepository()
    private let accountRegistries = AccountsRegistryRepository()

    func getFirstRegistryDate(completion: @escaping (Date) -> Void) {
        accountRegistries.getAll { regs in
            let first = regs.min(by: { $0.date < $1.date })?.date ?? Date()
            completion(first)
        }
    }

    func getAccountItems(params: TimelineFilterParams = TimelineFilterParams(), completion: @escaping ([TimelineDetailsItem]) -> Void) {
        // Load accounts and categories to allow local lookups
        accounts.getAll { accs in
            self.categories.getAll { _ in
                self.accountRegistries.getAll { regs in
                    let result = self.filterAndMap(regs: regs, accounts: accs, params: params)
                    completion(result)
                }
            }
        }
    }

    private func filterAndMap(regs: [AccountsRegistry], accounts: [Account], params: TimelineFilterParams) -> [TimelineDetailsItem] {
        let period = params.period
        let filtered = regs.filter { r in
            let paidOk = (params.paid == nil) || (params.paid == true && r.paid) || (params.paid == false && !r.paid)
            let periodOk = (period == nil) || (r.date >= period!.start && r.date <= period!.end)
            let categoryOk = params.categoryIds.isEmpty || (r.categoryId != nil && params.categoryIds.contains(r.categoryId!))
            let accountOk: Bool = {
                if params.accountIds.isEmpty {
                    // Without account filter, exclude archived accounts unless showArchived
                    let acc = accounts.first(where: { $0.id == r.accountId })
                    return params.showArchived || (acc?.includeInDash ?? true)
                } else {
                    return params.accountIds.contains(r.accountId)
                }
            }()
            return paidOk && periodOk && categoryOk && accountOk
        }

        let mapped: [TimelineDetailsItem] = filtered.map { reg in
            let accName = accounts.first(where: { $0.id == reg.accountId })?.name ?? "Unknown Source"
            return TimelineDetailsItem(
                sourceName: accName,
                registry: reg,
                category: categories.getById(reg.categoryId)
            )
        }

        let searched: [TimelineDetailsItem]
        if let q = params.search?.trimmingCharacters(in: .whitespacesAndNewlines), !q.isEmpty {
            searched = mapped.filter { d in
                (d.registry.description.lowercased().contains(q.lowercased()) || d.sourceName.lowercased().contains(q.lowercased()) || (d.category?.name.lowercased().contains(q.lowercased()) ?? false))
            }
        } else {
            searched = mapped
        }
        return searched.sorted { $0.registry.date > $1.registry.date }
    }
}

File: Web/src/data/service/TimelineService.ts
import { CreditCard, CreditCardInvoice, InvoiceTransaction, RegistryWithDetails } from "@models";
import { Repositories } from "@repositories";
import { getCurrentUser } from '../firebase/google-services';
import { AccountsRegistry } from '../models/AccountRegistry';
import FinancialMonthPeriod from "../utils/FinancialMonthPeriod";
import searchScore from "../utils/SearchScore";

export interface TimelineFilterPeriod {
  start: Date;
  end: Date;
}

export interface TimelineFilterParams {
  period?: TimelineFilterPeriod;
  categoryIds?: string[];
  showArchived?: boolean;
  accountIds?: string[];
  paid?: boolean;
  light?: boolean;
  search?: string;
}

export default class TimelineService {

  constructor(
    private repositories: Repositories,
    public readonly period = new FinancialMonthPeriod(1)
  ) {}

  public getFirstRegistryDate(): Date {
    const accountsRegistry = this.repositories.accountTransactions.firstRegistryDate;
    const invoicesRegistry = new Date() // this.repositories.creditCardsInvoices.firstRegistryDate;
    return accountsRegistry < invoicesRegistry ? accountsRegistry : invoicesRegistry;
  }

  public getAccountItems({
    period: periodData,
    categoryIds = [],
    accountIds = [],
    showArchived = false,
    light = false,
    paid,
    search
  }: TimelineFilterParams = {}): RegistryWithDetails[] {
    const { accounts, categories, accountTransactions: accountRegistries, creditCards, creditCardsInvoices } = this.repositories;
    const period = periodData && TimelineFilterPeriodImpl.fromData(periodData);

    const debit = accountRegistries.getCache()
      .filter(r => 
        // Paid
        (paid === undefined || r.paid === paid) &&
        // Period
        (!period || period.contains(r.date)) &&
        // Categories
        (categoryIds.length === 0 || (r.categoryId && categoryIds.includes(r.categoryId))) &&
        // Accounts
        (
          (accountIds.length === 0 && (showArchived || !accounts.getLocalById(r.accountId)?.archived)) ||
          accountIds.includes(r.accountId)
        )
      )
      .map<RegistryWithDetails>((registry) => ({
        registry,
        category: light ? undefined : categories.getLocalById(registry.categoryId),
        sourceName: accounts.getLocalById(registry.accountId)?.name || 'Unknown Source',
      }));

    const credit = categoryIds.length === 0 || categoryIds.includes(InvoiceTransaction.categoryId)
      ? creditCardsInvoices.getCache()
      .filter(r => 
        // Paid
        (paid === undefined || (paid ? r.paid : !r.paid)) &&
        // Period
        (!period || period.contains(r.paymentDate || r.invoiceDate)) &&
        // Accounts
        (
          accountIds.length === 0
          ? (showArchived || !accounts.getLocalById(this.findAccountIdOfInvoice(r))?.archived)
          : (accountIds.length > 0 && accountIds.includes(this.findAccountIdOfInvoice(r)))
        )
      )
      .map<RegistryWithDetails>((invoice) => ({
        registry: new InvoiceTransaction(
          invoice, 
          light ? new CreditCard(invoice.cardId) : creditCards.getLocalById(invoice.cardId)!
        ),
        category: !light ? categories.getLocalById(InvoiceTransaction.categoryId) : undefined,
        sourceName: accounts.getLocalById(invoice.paymentAccountId)?.name || 'Unknown Source',
      }))
      : [];

    const registries = ([...debit, ...credit])
      .sort(({registry: {date: a}}, {registry: {date: b}}) => b.getTime() - a.getTime());

    return search ? this.applySearch(registries, search) : registries;
  }

  private applySearch(registries: RegistryWithDetails[], query: string): RegistryWithDetails[] {
    const trimmed = query.trim();
    if (!trimmed) return registries;

    return registries
      .map(r => ({
        item: r,
        score: searchScore(trimmed, {
          description: r.registry.description,
          name: r.sourceName,
          category: r.category?.name,
          value: r.registry.value,
        }),
      }))
      .filter(r => r.score > 0)
      .sort((a, b) => b.score - a.score)
      .map(r => r.item);
  }

  private findAccountIdOfInvoice({ paymentAccountId, cardId}: CreditCardInvoice): string {
    if (paymentAccountId) return paymentAccountId;

    return this.repositories.creditCards
      .getLocalById(cardId)?.accountId
      || ''
    ;
  }
}

class TimelineFilterPeriodImpl {
  constructor(
    readonly start: Date,
    readonly end: Date
  ) {}

  public contains(date?: Date): boolean {
    if (!date) return false;
    return date >= this.start && date <= this.end;
  }

  static fromData(data: TimelineFilterPeriod) {
    return new TimelineFilterPeriodImpl(data.start, data.end);
  }
}
